File: ios/Runner/TuyaLogic/TuyaDeviceExtensions.swift
//
//  TuyaDeviceExtensions.swift
//  Runner
//
//  Created by Jacob Welin - Wejeo on 31/01/2023.
//

import Foundation


extension TuyaDeviceHandler: TuyaSmartActivatorDelegate {
    
    func activator(_ activator: TuyaSmartActivator!, didReceiveDevice deviceModel: TuyaSmartDeviceModel!, error: Error!) {
        print("Paring update...")
        
        if deviceModel != nil && error == nil {
            print("Device paired!")
            print("name: \(deviceModel.name ?? "Null"), name: \(deviceModel.uiName ?? "Null"), ")
            LocalDataHandler.currentDeviceId = deviceModel.devId
            self.initDevice(deviceId: deviceModel.devId)
            self.returnDevicePairResult(message: "Success:\(deviceModel.devId ?? "Null")")
        }
        
        if let e = error {
            print("Failed to pair device!")
            print("\(e)")
            self.returnDevicePairResult(message: "Failed to pair device")
        }
        
    }
}

extension TuyaDeviceHandler: TuyaSmartDeviceDelegate {
    
    func initDevice(deviceId:String){
        self.smartDevice = TuyaSmartDevice(deviceId: deviceId)
        self.smartDevice?.delegate = self
    }
    
    
    func updateDeviceData(device: TuyaSmartDevice) -> String {
        let data: [AnyHashable : Any] = [
            "name": device.deviceModel.name!,
            "devId": device.deviceModel.devId!,
            "isOnline": device.deviceModel.isOnline,
            "isCloudOnline": device.deviceModel.isCloudOnline,
            "onlineType": device.deviceModel.onlineType.rawValue,
            "deviceType": device.deviceModel.deviceType.rawValue,
            "dps": device.deviceModel.dps.toJSONString(),
            "homeId": device.deviceModel.homeId,
            "roomId": device.deviceModel.roomId,
        ]
        return data.toJSONString()
    }
    
    func device(_ device: TuyaSmartDevice, dpsUpdate dps: [AnyHashable : Any]) {
        print("Device dpsUpdate!")
        guard let eventSink = self.eventSink else{
            return
        }
        eventSink(updateDeviceData(device: device))
    }
    
//    func deviceOnlineUpdate(_ device: TuyaSmartDevice) {
//        print("deviceOnlineUpdate")
//        guard let eventSink = self.eventSink else{
//            return
//        }
//        eventSink(updateDeviceData(device: device))
//    }
    
    func deviceInfoUpdate(_ device: TuyaSmartDevice) {
        print("deviceInfoUpdate")
        guard let eventSink = self.eventSink else{
            return
        }
        eventSink(updateDeviceData(device: device))
    }
    
}


extension TuyaDeviceHandler: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        print("Device Stream started")
        
        self.eventSink = events
        guard let device = TuyaSmartDevice(deviceId: LocalDataHandler.currentDeviceId ?? "") else{
            print("Device Error: \(LocalDataHandler.currentDeviceId ?? "")")
            events(nil)
            return nil
        }
        
        let data: [AnyHashable : Any] = [
            "name": device.deviceModel.name!,
            "devId": device.deviceModel.devId!,
            "isOnline": device.deviceModel.isOnline,
            "isCloudOnline": device.deviceModel.isCloudOnline,
            "onlineType": device.deviceModel.onlineType.rawValue,
            "deviceType": device.deviceModel.deviceType.rawValue,
            "dps": device.deviceModel.dps.toJSONString(),
            "homeId": device.deviceModel.homeId,
            "roomId": device.deviceModel.roomId,
        ]
//        print(data.toJSONString())
        events(data.toJSONString())
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        print("Device Stream Canceled")
        return nil
    }
    
    
}

File: android/app/src/main/kotlin/dk/wejeo/wejeo_smart/TuyaLogic/TuyaDeviceExtensions.kt
package dk.wejeo.wejeo_smart.TuyaLogic

import android.util.Log
import com.tuya.smart.home.sdk.TuyaHomeSdk
import com.tuya.smart.sdk.api.IDevListener
import dk.wejeo.wejeo_smart.LocalDataHandler
import io.flutter.plugin.common.EventChannel
import org.json.JSONObject


class TuyaDeviceExtensions : EventChannel.StreamHandler, TuyaDeviceHandler() {
    val EVENT_LOG_TAG = "TuyaHomeEventStream"

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {

        print("Device Stream started")

        this.eventSink = events
        val currentDevId = LocalDataHandler.currentDeviceId
        if (currentDevId == null) {
            this.eventSink?.error("0", "No device id", "Could not get device id")
            return
        }
        updateDeviceData(currentDevId)
        startDeviceListener(currentDevId)

    }

    override fun onCancel(arguments: Any?) {
        this.eventSink = null
        Log.i(EVENT_LOG_TAG, "Device Stream Canceled")
    }

    private fun startDeviceListener(deviceId: String) {
        val mDevice = TuyaHomeSdk.newDeviceInstance(deviceId)
        mDevice.registerDevListener(object : IDevListener {
            override fun onDpUpdate(devId: String, dpStr: String) {
                Log.i(EVENT_LOG_TAG, "onDpUpdate")
            }
            override fun onRemoved(devId: String) {
                Log.i(EVENT_LOG_TAG, "onRemoved")
            }
            override fun onStatusChanged(devId: String, online: Boolean) {
                Log.i(EVENT_LOG_TAG, "onStatusChanged")
            }
            override fun onNetworkStatusChanged(devId: String, status: Boolean) {
                Log.i(EVENT_LOG_TAG, "onNetworkStatusChanged")
            }
            override fun onDevInfoUpdate(devId: String) {
                Log.i(EVENT_LOG_TAG, "onDevInfoUpdate")
            }
        })
    }

    private fun updateDeviceData(deviceId: String) {
        val device = TuyaHomeSdk.getDataInstance().getDeviceBean(deviceId)
        if (device == null) {
            this.eventSink?.error("0", "Device null", "Could not get device from id")
            return
        }
        val data: Map<String, Any?> =
            mapOf(
                "name" to device.name,
                "devId" to device.devId,
                "isOnline" to device.isOnline,
                "isCloudOnline" to device.isCloudOnline,
                "onlineType" to 0,
                "deviceType" to 0,
                "dps" to JSONObject(device.dps).toString(),
                "homeId" to 0,
                "roomId" to 0,
            )
        val jsData = JSONObject(data)
        eventSink?.success(jsData.toString())
    }

}